### PR TITLE
Last bits of WeaponSet

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/object/behavior/spawnbehavior.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        AbleToAttackType type, const Object *target, const Coord3D *location, CommandSourceType source) = 0;
+    virtual bool Can_Any_Slaves_Attack() = 0;
+    virtual void Order_Slaves_To_Go_Idle(CommandSourceType source) = 0;
+    virtual void Order_Slaves_Disabled_Until(DisabledType type, unsigned int frale) = 0;
+    virtual void Order_Slaves_To_Clear_Disabled(DisabledType type) = 0;
+    virtual void Give_Slaves_Stealth_Upgrade(bool stealth) = 0;
+    virtual bool Are_All_Slaves_Stealthed() const = 0;
+    virtual void Reveal_Slaves() = 0;
+    virtual bool Do_Slaves_Have_Freedom() const = 0;
+};
```
Comment (xezon) [inline src/game/logic/object/behavior/spawnbehavior.h]: eol

File: src/game/logic/object/weaponset.cpp [hunk 2]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
     return m_weapons[wslot];
 }
 
-#if 0
-bool WeaponSet::Choose_Best_Weapon_For_Target(
-    const Object *source_obj, const Object *victim_obj, WeaponChoiceCriteria criteria, CommandSourceType source)
+bool Is_Attack_Type_1(AbleToAttackType type)
 {
+    return (type & ATTACK_TYPE_1) != 0;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp:420]: Are these names
```
    ATTACK_RESULT_0 = 0,
    ATTACK_RESULT_1,
    ATTACK_RESULT_2,
    ATTACK_RESULT_3,

    ATTACK_TYPE_0 = 0,
    ATTACK_TYPE_1,
    ATTACK_TYPE_2,
    ATTACK_TYPE_3,
    ATTACK_TYPE_4,
```
the real official names or do we miss proper names yet?

File: src/game/logic/object/weaponset.cpp [hunk 3]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+        }
+    }
+
+    return Get_Able_To_Use_Weapon_Against_Target(type, source_obj, victim_obj, victim_obj->Get_Position(), source, wslot);
+}
+
+int Get_Victim_Anti_Mask(const Object *victim)
+{
+    if (victim->Is_KindOf(KINDOF_SMALL_MISSILE)) {
+        return 8;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp:529]: Do we have enum values for these bit numbers?

File: src/game/logic/object/weaponset.cpp [hunk 4]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+        StealthUpdate *update = victim_obj->Get_Stealth_Update();
+
+        if (update != nullptr) {
+            if (update->Has_Disguised_Template()) {
+                Player *player = source_obj->Get_Controlling_Player();
+                Player *player2 = g_thePlayerList->Get_Nth_Player(update->Get_Player_Index());
+
+                if (player != nullptr) {
+                    if (player2 != nullptr) {
+                        if (player->Get_Relationship(player2->Get_Default_Team())) {
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: This condition should compare with relationship value
```
enum Relationship
{
    ENEMIES,
    NEUTRAL,
    ALLIES,
};
```

File: src/game/logic/object/weaponset.cpp [hunk 5]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+            start_slot = m_curWeapon;
+            end_slot = m_curWeapon;
+        } else if (wslot == WEAPONSLOT_UNK) {
+            start_slot = WEAPONSLOT_TERTIARY;
+            end_slot = WEAPONSLOT_PRIMARY;
+        } else {
+            start_slot = wslot;
+            end_slot = wslot;
+        }
+
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp:654]:
```
static_assert(WEAPONSLOT_PRIMARY <= WEAPONSLOT_TERTIARY)
```
perhaps

File: src/game/logic/object/weaponset.cpp [hunk 6]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+            if (list != nullptr) {
+                for (auto it = list->begin(); it != list->end(); it++) {
+                    Object *obj = *it;
+
+                    if (obj->Is_Able_To_Attack()) {
+                        CanAttackResult can =
+                            obj->Get_Able_To_Use_Weapon_Against_Target(type, victim_obj, location, source, WEAPONSLOT_UNK);
+
+                        if (can == ATTACK_RESULT_3 || can == ATTACK_RESULT_2) {
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: Perhaps swap to make it easier on the brain
```
can == ATTACK_RESULT_2 || can == ATTACK_RESULT_3
```

File: src/game/logic/object/weaponset.cpp [hunk 7]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+        }
+    }
+
+    if ((source_obj->Is_KindOf(KINDOF_IMMOBILE) || source_obj->Is_KindOf(KINDOF_SPAWNS_ARE_THE_WEAPONS)
+            || containedby != nullptr)
+        && has_weapon && !weapon_in_range && type != ATTACK_TYPE_4) {
+        return ATTACK_RESULT_1;
+    }
+
+    CanAttackResult result = static_cast<CanAttackResult>(is_in_range + 2);
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: Is this a compiler optimization for 
```
CanAttackResult result = is_in_range ? ATTACK_RESULT_3 : ATTACK_RESULT_2
```
?

File: src/game/logic/object/weaponset.cpp [hunk 8]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+    if (victim_obj != nullptr) {
+        bool b1 = false;
+        bool b2 = false;
+        float f1 = 0.0f;
+        float f2 = 0.0f;
+        float f3 = 0.0f;
+        float f4 = 0.0f;
+        int i1 = 0;
+        int i2 = 0;
+
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp:720]:
```
static_assert(WEAPONSLOT_TERTIARY >= 0)
```
perhaps

File: src/game/logic/object/weaponset.cpp [hunk 9]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+                if (((1 << source) & mask) != 0 || (mask & 4) != 0) {
+                    Weapon *weapon = m_weapons[i];
+
+                    if (weapon != nullptr
+                        && (weapon->Get_Status() != Weapon::OUT_OF_AMMO || weapon->Is_Auto_Reloads_Clip())) {
+                        if ((Get_Victim_Anti_Mask(victim_obj) & weapon->Get_Anti_Mask()) != 0
+                            && weapon->Is_Within_Target_Pitch(source_obj, victim_obj)) {
+                            float damage = weapon->Estimate_Weapon_Damage(source_obj, victim_obj);
+                            float range = weapon->Get_Attack_Range(source_obj);
+                            bool b3 = weapon->Get_Status() == Weapon::READY_TO_FIRE;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: can_attack

File: src/game/logic/object/weaponset.cpp [hunk 10]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+                            }
+
+                            if (damage > 0.0f || weapon->Get_Damage_Type() == DAMAGE_UNRESISTABLE) {
+                                BitFlags<KINDOF_COUNT> against =
+                                    m_curWeaponTemplateSet->Get_Nth_Preferred_Against_Mask(static_cast<WeaponSlotType>(i));
+
+                                if (against.Any() && victim_obj->Is_KindOf_Multi(against, KINDOFMASK_NONE)) {
+                                    damage = 10000000000.0f;
+                                    range = 10000000000.0f;
+                                    b3 = weapon->Get_Status() != Weapon::OUT_OF_AMMO;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: Is this correct? b3 is just overwritten.

File: src/game/logic/object/weaponset.cpp [hunk 11]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+
+    if (victim_obj != nullptr) {
+        bool b1 = false;
+        bool b2 = false;
+        float f1 = 0.0f;
+        float f2 = 0.0f;
+        float f3 = 0.0f;
+        float f4 = 0.0f;
+        int i1 = 0;
+        int i2 = 0;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: free_weapon
busy_weapon

Can be `WeaponSlotType`

File: src/game/logic/object/weaponset.cpp [hunk 12]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+bool WeaponSet::Choose_Best_Weapon_For_Target(
+    const Object *source_obj, const Object *victim_obj, WeaponChoiceCriteria criteria, CommandSourceType source)
+{
+    if (Is_Cur_Weapon_Locked()) {
+        return true;
+    }
+
+    if (victim_obj != nullptr) {
+        bool b1 = false;
+        bool b2 = false;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: use_free_weapon
use_busy_weapon

File: src/game/logic/object/weaponset.cpp [hunk 13]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+        if (b1) {
+            m_curWeapon = static_cast<WeaponSlotType>(i1);
+        } else if (b2) {
+            m_curWeapon = static_cast<WeaponSlotType>(i2);
+            return true;
+        } else {
+            m_curWeapon = WEAPONSLOT_PRIMARY;
+        }
+
+        return b1;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: Maybe instead write
```c++
        if (b1) {
            m_curWeapon = static_cast<WeaponSlotType>(i1);
            return true;
        } else if (b2) {
            m_curWeapon = static_cast<WeaponSlotType>(i2);
            return true;
        } else {
            m_curWeapon = WEAPONSLOT_PRIMARY;
            return false;
        }
```

File: src/game/logic/object/weaponset.cpp [hunk 14]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+        return true;
+    }
+
+    if (victim_obj != nullptr) {
+        bool b1 = false;
+        bool b2 = false;
+        float f1 = 0.0f;
+        float f2 = 0.0f;
+        float f3 = 0.0f;
+        float f4 = 0.0f;
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp]: free_weapon_max_range
free_weapon_max_damage
busy_weapon_max_range
busy_weapon_max_damage

File: src/game/logic/object/weaponset.cpp [hunk 15]
```diff
@@ -409,23 +415,385 @@ Weapon *WeaponSet::Get_Weapon_In_Weapon_Slot(WeaponSlotType wslot) const
...
+                                    m_curWeaponTemplateSet->Get_Nth_Preferred_Against_Mask(static_cast<WeaponSlotType>(i));
+
+                                if (against.Any() && victim_obj->Is_KindOf_Multi(against, KINDOFMASK_NONE)) {
+                                    damage = 10000000000.0f;
+                                    range = 10000000000.0f;
+                                    b3 = weapon->Get_Status() != Weapon::OUT_OF_AMMO;
+                                }
+
+                                if (criteria != PREFER_MOST_DAMAGE) {
+                                    if (criteria == PREFER_LONGEST_RANGE) {
```
Comment (xezon) [inline src/game/logic/object/weaponset.cpp:754]: `WeaponChoiceCriteria` only has 2 states. Why is there a double test if this has no `else` ?

File: src/game/logic/object/behavior/spawnbehavior.h (hunk 1 above)
Comment (jonwil) [reply src/game/logic/object/behavior/spawnbehavior.h]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp:420]: We don't have correct names yet.

File: src/game/logic/object/weaponset.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp:529]: We don't.

File: src/game/logic/object/weaponset.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: Fixed.

File: src/game/logic/object/weaponset.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: Yes its correct.

File: src/game/logic/object/weaponset.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp]: fixed.

File: src/game/logic/object/weaponset.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/logic/object/weaponset.cpp:754]: No idea.

**Review by xezon — APPROVED** [review]
I tested in Release configuration. Shell Map looked as usual. Test map with units attacking practice targets looked as usual.

**Pull request merged**